File: Ash/OS_LAB/7_rwsemaphore.c
#include <stdio.h>
#include <pthread.h>
#include <semaphore.h>

sem_t db;
sem_t x;
int readerCount = 0;

void *reader(void *arg)
{

    sem_wait(&x);
    readerCount++;
    printf("Reader %ld entered db\n", (long)arg);
    if (readerCount == 1)
    {
        sem_wait(&db);
    }
    printf("Reader %ld is reading...\n", (long)arg);

    sem_post(&x);

    sem_wait(&x);
    readerCount--;
    printf("Reader %ld exited from db\n", (long)arg);
    if (readerCount == 0)
    {
        sem_post(&db);
    }
    sem_post(&x);

    pthread_exit(NULL);
}

void *writer(void *arg)
{

    sem_wait(&db);
    printf("Writer %ld entered db\n", (long)arg);
    printf("Writer %ld is writing...\n", (long)arg);
    sem_post(&db);
    printf("Writer %ld exited from db\n", (long)arg);

    pthread_exit(NULL);
}

int main()
{
    pthread_t readers[5], writers[5];
    sem_init(&db, 0, 1);
    sem_init(&x, 0, 1);

    for (long i = 0; i < 5; i++)
    {
        pthread_create(&readers[i], NULL, reader, (void *)i);
        pthread_create(&writers[i], NULL, writer, (void *)i);
    }

    for (int i = 0; i < 5; i++)
    {
        pthread_join(readers[i], NULL);
        pthread_join(writers[i], NULL);
    }

    sem_destroy(&db);
    sem_destroy(&x);

    return 0;
}

File: Ash/OS_LAB/7_rwsemaphore.py
import threading
import time

db = threading.Semaphore(1)
x = threading.Semaphore(1)
readerCount = 0

def reader(arg):
    global readerCount

    x.acquire()
    readerCount += 1
    print(f"Reader {arg} entered db")
    if readerCount == 1:
        db.acquire()
    print(f"Reader {arg} is reading...")
    x.release()

    time.sleep(1)  # Simulate reading time

    x.acquire()
    readerCount -= 1
    print(f"Reader {arg} exited from db")
    if readerCount == 0:
        db.release()
    x.release()

def writer(arg):
    db.acquire()
    print(f"Writer {arg} entered db")
    print(f"Writer {arg} is writing...")
    time.sleep(1)  # Simulate writing time
    db.release()
    print(f"Writer {arg} exited from db")

if __name__ == "__main__":
    readers = []
    writers = []

    for i in range(5):
        r = threading.Thread(target=reader, args=(i,))
        w = threading.Thread(target=writer, args=(i,))
        readers.append(r)
        writers.append(w)
        r.start()
        w.start()

    for r in readers:
        r.join()
    for w in writers:
        w.join()


